docs: config (#495)
libp2p/js-libp2p: doc/CONFIGURATION.md

# Configuration

* [Overview](#overview)
* [Modules](#modules)
  * [Transport](#transport)
  * [Stream Multiplexing](#stream-multiplexing)
  * [Connection Encryption](#connection-encryption)
  * [Peer Discovery](#peer-discovery)
  * [Content Routing](#content-routing)
  * [Peer Routing](#peer-routing)
  * [DHT](#dht)
  * [Pubsub](#pubsub)
* [Libp2p node](#libp2p-node)
* [Configuration examples](#configuration-examples)

## Overview

> Considering libp2p's modular nature, it can be composed by a set of building blocks that enable users to create any type of p2p systems without needing to implement its network layer. Taking into account that each p2p system has its own needs and requirements, libp2p can be configured accordingly.

Reviewer: ```suggestion
libp2p is a modular networking stack. It's designed to be able to suit a variety of project needs. The configuration of libp2p is a key part of its structure. It enables you to bring exactly what you need, and only what you need. This document is a guide on how to configure libp2p for your particular project. Check out the [Configuration examples](#configuration-examples) section if you're just looking to leverage an existing configuration.
```

> `js-libp2p` acts as the composer for this modular p2p networking stack using libp2p compatible modules as its building blocks.

Reviewer: I think this is superfluous. 
```suggestion
```

Author: Maybe I should move this under the next section? Its main intend is to introduce the `modules`

> While libp2p building blocks may vary, `js-libp2p` top level API will be consistent.

Reviewer: ```suggestion
Regardless of how you configure libp2p, the top level [API](./API.md) will always remain the same. **Note**: if some modules are not configured, like Content Routing, using those methods will throw errors.
```

## Modules

> For getting an instance of `js-libp2p` compliant with all types of networking requirements, it is possible to specify the following building blocks:

Reviewer: I think using `subsystems` instead of `building blocks` is a bit more descriptive.

- https://github.com/libp2p/specs/blob/master/discovery/mdns.md

### Content Routing

> Content routing provides a way to find where content lives in the network. It works in two steps: 1) Peers provide (announce) to the network that they are holders of specific content and 2) Peers issue queries to find where that content lives. A Content Routing mechanism could be as complex as a DHT or a simple registry somewhere in the network.

The content routing modules available are:

- [js-libp2p-kad-dht](https://github.com/libp2p/js-libp2p-kad-dht)
- [js-libp2p-delegated-peer-routing](https://github.com/libp2p/js-libp2p-delegated-peer-routing)

If none of the available content routing protocols fulfills your needs, you can create a libp2p compatible one. A libp2p content routing protocol just needs to be compliant with the [interface-content-routing](https://github.com/libp2p/js-interfaces/tree/master/src/content-routing). **(WIP: This module is not yet implemented)**

If you want to get to know more about libp2p content routing, you should read the following content:

- https://docs.libp2p.io/concepts/content-routing

### Peer Routing

> > Peer Routing offers a way to find other peers in the network by intentionally issuing queries, iterative or recursive, until a Peer is found or the closest Peers, given the Peer Routing algorithm strategy, are found.

Reviewer 2: ```suggestion
> Peer Routing offers a way to find other peers in the network by issuing queries using a Peer Routing algorithm, which may be iterative or recursive. If the algorithm is unable to find the target peer, it will return the peers that are "closest" to the target peer, using a distance metric defined by the algorithm.
```

The pubsub routers available are:

- [libp2p/js-libp2p-floodsub](https://github.com/libp2p/js-libp2p-floodsub)
- [ChainSafe/gossipsub-js](https://github.com/ChainSafe/gossipsub-js)

If none of the available pubsub routers fulfills your needs, you can create a libp2p compatible one. A libp2p pubsub router just needs to be created on top of [libp2p/js-libp2p-pubsub](https://github.com/libp2p/js-libp2p-pubsub), which ensures `js-libp2p` API expectations.

If you want to get to know more about libp2p pubsub, you should read the following content:

- https://docs.libp2p.io/concepts/publish-subscribe
- https://github.com/libp2p/specs/tree/master/pubsub

## Libp2p node

When [creating a libp2p node](./API.md#create), the modules needed should be specified in the following way:

```js
const modules = {
  transport: [],
  streamMuxer: [],
  connEncryption: [],
  contentRouting: [],
  peerRouting: [],
  peerDiscovery: [],
  dht: dhtImplementation,
  pubsub: pubsubImplementation
}
```

> Moreover, the majority of the modules can be customized via option parameters. This way, it is also possible to provide this options through a config object. This config object should have the property name of each building block to configure, the same way as the modules specification.

Author: Yeah, I agree with you! I will break this in smaller pieces

Regardless of how you configure libp2p, the top level [API](./API.md) will always remain the same. **Note**: if some modules are not configured, like Content Routing, using those methods will throw errors.

## Modules

`js-libp2p` acts as the composer for this modular p2p networking stack using libp2p compatible modules as its subsystems. For getting an instance of `js-libp2p` compliant with all types of networking requirements, it is possible to specify the following subsystems:

- Transports
- Multiplexers
- Connection encryption mechanisms
- Peer discovery protocols
- Content routing protocols
- Peer routing protocols
- DHT implementation
- Pubsub router

The libp2p ecosystem contains at least one module for each of these subsystems. The user should install and import the modules that are relevant for their requirements. Moreover, thanks to the existing interfaces it is easy to create a libp2p compatible module and use it.

After selecting the modules to use, it is also possible to configure each one according to your needs.

> Bear in mind that only a **transport** and a **connection encryption** are required, while all the other subsystems are optional.

Reviewer: ```suggestion
Bear in mind that only a **transport** and **connection encryption** are required, while all the other subsystems are optional.
```
